internal/util/reftracker/reftracker.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reftracker
+
+import (
+	goerrors "errors"
+	"fmt"
+
+	"github.com/ceph/go-ceph/rados"
+
+	"github.com/ceph/ceph-csi/internal/util/reftracker/errors"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/radoswrapper"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/reftype"
+	v1 "github.com/ceph/ceph-csi/internal/util/reftracker/v1"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/version"
+)
+
+// reftracker is key-based implementation of a reference counter.
+//
+// Unlike integer-based counter, reftracker counts references by tracking
+// unique keys. This allows accounting in situations where idempotency must be
+// preserved. It guarantees there will be no duplicit increments or decrements
+// of the counter.
+//
+// It is stored persistently as a RADOS object, and is safe to use with
+// multiple concurrent writers, and across different nodes of a cluster.
+
+// Add atomically adds refs to reference tracker named `rtName`.
+// If the reftracker object doesn't exist yet, it is created. This is also
+// reflected in the return value.
+func Add(
+	ioctx radoswrapper.IOContextW,
+	rtName string,
+	refs map[string]struct{},
+) (created bool, err error) {
+	if err = validateAddInput(rtName, refs); err != nil {
+		return false, err
+	}
+
+	// Read reftracker version.
+
```

Comment:
no new line required

Reply:
Comments separated with blank lines above and below are describing sections of code. Without having a new line separating it from the code, it would look like it's describing only that particular line it was adjacent to. I tried to stay consistent with this rule throughout this PR.

I'm leaving it like this for now, please let me know if you insist on changing it.